lib/base/bulker.hpp

```diff
@@ -0,0 +1,116 @@
+/* Icinga 2 | (c) 2022 Icinga GmbH | GPLv2+ */
+
+#ifndef BULKER_H
+#define BULKER_H
+
+#include <boost/config/detail/suffix.hpp>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <utility>
+#include <vector>
+
+namespace icinga
+{
+
+/**
+ * A queue which outputs the input as bulks of a defined size
+ * or after a defined time, whichever is reached first
+ *
+ * @ingroup base
+ */
+template<class T>
+class Bulker
+{
+private:
+	typedef std::chrono::steady_clock Clock;
+
+public:
+	typedef std::vector<T> Container;
+	typedef typename Container::size_type SizeType;
+	typedef typename Clock::duration Duration;
+
+	Bulker(SizeType bulkSize, Duration threshold)
+		: m_BulkSize(bulkSize), m_Threshold(threshold), m_NextConsumption(NullTimePoint()) { }
+
+	void ProduceOne(T needle);
+	Container ConsumeMany();
+
+private:
+	typedef std::chrono::time_point<Clock> TimePoint;
+
+	static inline
+	TimePoint NullTimePoint()
+	{
+		return TimePoint::min();
+	}
+
+	inline void UpdateNextConsumption()
+	{
+		m_NextConsumption = Clock::now() + m_Threshold;
+	}
+
+	const SizeType m_BulkSize;
+	const Duration m_Threshold;
+
+	std::mutex m_Mutex;
+	std::condition_variable m_CV;
+	std::queue<Container> m_Bulks;
+	TimePoint m_NextConsumption;
+};
+
+template<class T>
+void Bulker<T>::ProduceOne(T needle)
+{
+	std::unique_lock<std::mutex> lock (m_Mutex);
+
+	if (m_Bulks.empty() || m_Bulks.back().size() == m_BulkSize) {
+		m_Bulks.emplace();
+	}
+
+	m_Bulks.back().emplace_back(std::move(needle));
+
+	if (m_Bulks.size() == 1u && m_Bulks.back().size() == m_BulkSize) {
+		m_CV.notify_one();
+	}
+}
+
+template<class T>
+typename Bulker<T>::Container Bulker<T>::ConsumeMany()
+{
+	std::unique_lock<std::mutex> lock (m_Mutex);
+
+	if (BOOST_UNLIKELY(m_NextConsumption == NullTimePoint())) {
+		UpdateNextConsumption();
+	}
+
+	for (;;) {
+		if (!m_Bulks.empty() && m_Bulks.front().size() == m_BulkSize) {
+			auto haystack (std::move(m_Bulks.front()));
+
+			m_Bulks.pop();
+			UpdateNextConsumption();
+			return std::move(haystack);
+		}
+
+		auto deadline (m_NextConsumption);
+
+		if (m_CV.wait_until(lock, deadline) == std::cv_status::timeout) {
```

Comment:
No. The waiting could be also woken up sporadically, w/o any actual reason. In the worst case we'd wait 249ms, another 249ms, another 249ms... Even if we wouldn’t take this into account, I want to sleep at most not **for 250ms**, but for 250ms **since the last consumption**.

Reply:
> I want to sleep at most not **for 250ms**, but for 250ms **since the last consumption**.

Is there any particular reason for this? I don't think that such details are relevant for the function we need. On second thought: Why would you even want to wake a caller periodically even if no items are returned?

I think for our use-case it should even be fine to return everything immediately once anything is available, even if this results in chunks of size 1. As soon as there's enough throughput that we can no longer handle chunks of size 1 anymore, the chunks will get bigger automatically.

Reply:
1. It just feels right, more than all alternatives.
2. To get a chance to break the loop.
3. IMAO that's not what a bulker is for.

Reply:
2: Good point, overlooked that.

Apart from that, we are not adding a bulker for the sake of adding a bulker but rather because we want to avoid a latency penalty. I think you could achieve this goal with simpler code and we don't have any plans for using this for anything else. I won't insist on this, this was just a suggestion how I think things could be simplified.

Reply:
IMAO a generic.. thing shall operate.. generically.

Reply:
Was building this specific generic thing ever an requirement? I'd be happy with anything that gets the job done, it wouldn't have to be called bulker.

Reply:
No. But it's best practice.